python/glow/wgr/linear_model/ridge_model.py

```diff
@@ -19,153 +19,212 @@
 warnings.filterwarnings('ignore', category=UserWarning, message='.*applyInPandas.*')
 
-__all__ = ['RidgeReducer', 'RidgeRegression']
+__all__ = ['RidgeReduction', 'RidgeRegression']
 
 
 @typechecked
-class RidgeReducer:
+class RidgeReduction:
     """
     The RidgeReducer class is intended to reduce the feature space of an N by M block matrix X to an N by P<<M block
     matrix.  This is done by fitting K ridge models within each block of X on one or more target labels, such that a
     block with L columns to begin with will be reduced to a block with K columns, where each column is the prediction
     of one ridge model for one target label.
     """
-    def __init__(self, alphas: NDArray[(Any, ), Float] = np.array([])) -> None:
+    def __init__(self,
+                 block_df: DataFrame,
+                 label_df: pd.DataFrame,
+                 sample_blocks: Dict[str, List[str]],
+                 cov_df: pd.DataFrame = pd.DataFrame({}),
+                 alphas: NDArray[(Any, ), Float] = np.array([]),
+                 label_type='detect') -> None:
         """
-        RidgeReducer is initialized with a list of alpha values.
-
         Args:
+            block_df : Spark DataFrame representing the beginning block matrix X
+            label_df : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            cov_df : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble (optional).
             alphas : array_like of alpha values used in the ridge reduction (optional).
+            label_type:
         """
-        if not (alphas >= 0).all():
-            raise Exception('Alpha values must all be non-negative.')
-        self.alphas = create_alpha_dict(alphas)
-
-    def fit(
-        self,
-        blockdf: DataFrame,
-        labeldf: pd.DataFrame,
-        sample_blocks: Dict[str, List[str]],
-        covdf: pd.DataFrame = pd.DataFrame({})) -> DataFrame:
+        self.set_block_df(block_df)
+        self.set_sample_blocks(sample_blocks)
+        self.set_label_type(label_type)
+        self.set_label_df(label_df)
+        self.set_cov_df(cov_df)
+        self.set_alphas(alphas)
+        self.set_model_df()
+        self.set_reduced_block_df()
+
+    def __getstate__(self):
+        # Copy the object's state from self.__dict__ which contains
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state['_RidgeReduction__block_df'], state['_RidgeReduction__model_df'], state[
+            '_RidgeReduction__reduced_block_df']
+        return state
+
+    def __setstate__(self, state):
+        # Restore instance attributes
+        self.__dict__.update(state)
+        self.__block_df = [x for x in globals().values() if id(x) == self.__block_df_id]
```

Comment:
Actually, why do we even need this `__setstate__`? When is it called on the driver? I'm not sure we need these ids to be picklable.

Reply:
This is somewhat based on what you gave me (https://stackoverflow.com/questions/2345944/exclude-objects-field-from-pickling-in-python). As far as I undertand, the essence is to keep a picklable surrogate for the object, drop object from pickling in `__getstate__` and then revive the object inside the class in ``__setstate__`` using the surrogate. Here I am using the object id as the surrogate because I am dealing with a variable. In the example in the link, since it is dealing with a file, it is using filename. On my local tests this works and without it spark dfs do not pickle. Do you think there may be a problem with this on a real cluster?

Reply:
You are right. removed variable recovery from setstate